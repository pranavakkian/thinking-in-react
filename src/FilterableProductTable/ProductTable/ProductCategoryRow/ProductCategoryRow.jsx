import PropTypes from 'prop-types';
import React, { memo } from 'react';

const ProductCategoryRow = (props) => {
  const { category } = props;
  return (
    <th id="categoryHeader" colSpan="2">
      {category}
    </th>
  );
};

ProductCategoryRow.propTypes = {
  category: PropTypes.string.isRequired,
};

export default memo(ProductCategoryRow);
