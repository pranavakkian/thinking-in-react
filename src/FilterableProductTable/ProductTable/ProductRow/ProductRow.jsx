import PropTypes from 'prop-types';
import React, { memo } from 'react';

const ProductRow = (props) => {
  const { data } = props;
  const { name, price } = data;

  return (
    <>
      <td id="dataName">{name}</td>
      <td id="dataPrice">{price}</td>
    </>
  );
};

ProductRow.propTypes = {
  data: PropTypes.shape({
    category: PropTypes.string.isRequired,
    price: PropTypes.string.isRequired,
    stocked: PropTypes.bool.isRequired,
    name: PropTypes.string.isRequired,
  }).isRequired,
};

export default memo(ProductRow);
