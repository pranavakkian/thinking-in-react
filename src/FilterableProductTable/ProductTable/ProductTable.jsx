import PropTypes from 'prop-types';
import React, { Fragment, memo } from 'react';
import ProductCategoryRow from './ProductCategoryRow/ProductCategoryRow';
import ProductRow from './ProductRow/ProductRow';

const ProductTable = (props) => {
  const { data } = props;
  const headingSet = new Set();
  data.map((item) => headingSet.add(item.category));
  const headings = [];
  headingSet.forEach((item) => headings.push(item));

  return (
    <table>
      <thead>
        <tr>
          <th id="nameHeader">Name</th>
          <th id="priceHeader">Price</th>
        </tr>
      </thead>
      <tbody>
        {headings.map((heading) => (
          <Fragment key={heading}>
            <tr>
              <ProductCategoryRow category={heading} />
            </tr>
            {data
              .filter((item) => item.category === heading)
              .map((value) => (
                <tr key={value.name}>
                  <ProductRow data={value} />
                </tr>
              ))}
          </Fragment>
        ))}
      </tbody>
    </table>
  );
};

ProductTable.propTypes = {
  data: PropTypes.arrayOf(
    PropTypes.shape({
      category: PropTypes.string.isRequired,
      price: PropTypes.string.isRequired,
      stocked: PropTypes.bool.isRequired,
      name: PropTypes.string.isRequired,
    }),
  ).isRequired,
};

export default memo(ProductTable);
