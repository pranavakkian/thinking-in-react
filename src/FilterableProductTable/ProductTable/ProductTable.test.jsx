/* eslint-disable no-undef */
import { mount } from 'enzyme';
import React from 'react';
import ProductCategoryRow from './ProductCategoryRow/ProductCategoryRow';
import ProductRow from './ProductRow/ProductRow';
import ProductTable from './ProductTable';

const data = [
  {
    category: 'Sporting Goods',
    price: '$29.99',
    stocked: false,
    name: 'Basketball',
  },
  {
    category: 'Electronics',
    price: '$99.99',
    stocked: true,
    name: 'iPod Touch',
  },
];

const wrapper = mount(<ProductTable data={data} />);

describe('ProductTable', () => {
  it('ProductTable rendered successfully!', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Number of each table elements!', () => {
    expect(wrapper.find('table')).toHaveLength(1);
    expect(wrapper.find('thead')).toHaveLength(1);
    expect(wrapper.find('tbody')).toHaveLength(1);
    expect(wrapper.find('th')).toHaveLength(4);
    expect(wrapper.find('tr')).toHaveLength(5);
    expect(wrapper.find('td')).toHaveLength(4);
  });

  it('Table header text content!', () => {
    expect(wrapper.find('#nameHeader').text()).toEqual('Name');
    expect(wrapper.find('#priceHeader').text()).toEqual('Price');
  });

  it('Child components rendered correctly!', () => {
    expect(
      wrapper.contains(<ProductCategoryRow category="Electronics" />),
    ).toBeTruthy();
    expect(wrapper.contains(<ProductRow data={data[0]} />)).toBeTruthy();
  });

  it('ProductTable un-mounted successfully!', () => {
    wrapper.unmount();
  });
});
