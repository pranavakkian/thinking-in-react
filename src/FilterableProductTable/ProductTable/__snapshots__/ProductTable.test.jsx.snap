// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTable ProductTable rendered successfully! 1`] = `
<Memo(ProductTable)
  data={
    Array [
      Object {
        "category": "Sporting Goods",
        "name": "Basketball",
        "price": "$29.99",
        "stocked": false,
      },
      Object {
        "category": "Electronics",
        "name": "iPod Touch",
        "price": "$99.99",
        "stocked": true,
      },
    ]
  }
>
  <table>
    <thead>
      <tr>
        <th
          id="nameHeader"
        >
          Name
        </th>
        <th
          id="priceHeader"
        >
          Price
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <Memo(ProductCategoryRow)
          category="Sporting Goods"
        >
          <th
            colSpan="2"
            id="categoryHeader"
          >
            Sporting Goods
          </th>
        </Memo(ProductCategoryRow)>
      </tr>
      <tr
        key="Basketball"
      >
        <Memo(ProductRow)
          data={
            Object {
              "category": "Sporting Goods",
              "name": "Basketball",
              "price": "$29.99",
              "stocked": false,
            }
          }
        >
          <td
            id="dataName"
          >
            Basketball
          </td>
          <td
            id="dataPrice"
          >
            $29.99
          </td>
        </Memo(ProductRow)>
      </tr>
      <tr>
        <Memo(ProductCategoryRow)
          category="Electronics"
        >
          <th
            colSpan="2"
            id="categoryHeader"
          >
            Electronics
          </th>
        </Memo(ProductCategoryRow)>
      </tr>
      <tr
        key="iPod Touch"
      >
        <Memo(ProductRow)
          data={
            Object {
              "category": "Electronics",
              "name": "iPod Touch",
              "price": "$99.99",
              "stocked": true,
            }
          }
        >
          <td
            id="dataName"
          >
            iPod Touch
          </td>
          <td
            id="dataPrice"
          >
            $99.99
          </td>
        </Memo(ProductRow)>
      </tr>
    </tbody>
  </table>
</Memo(ProductTable)>
`;
